src/peft/pipelines/base.py

```diff
@@ -0,0 +1,71 @@
+# coding=utf-8
+# Copyright 2023-present the HuggingFace Inc. team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from abc import ABC, abstractmethod
+from typing import Any, Union
+
+import torch
+
+
+class BasePeftPipeline(ABC):
+    r"""
+    Base skeleton class for all PEFT pipelines. This class cannot be used as it is and should be subclassed and the
+    method `_load_model_and_processor` should be implemented.
+
+    Attributes:
+        transformers_model_class (`transformers.PreTrainedModel`):
+            The class of the transformers model to use.
+        transformers_processor_class (Union[`transformers.PreTrainedTokenizer`, `transformers.AutoProcessor`]):
+            The class of the transformers processor to use.
+        task_type (str):
+            The task type of the pipeline - i.e. the name of the pipeline.
+        device (Union[str, int, torch.device]):
+            The device to run the pipeline on.
+        supported_extra_args (tuple):
+            The list of extra arguments supported by the pipeline.
+    """
+    transformers_model_class = None
```

Comment:
For my better understanding, why are these attributes defined as class attributes and not inside of `__init__`?

Reply:
I did that to not bloat the `__init__` method and pass to that method only useful attributes such as the model, processor (tokenizer), etc. 
I thought it would be then easier in the future for contributors to add pipelines by just overriding these classes in the child class and not let them modify these values by passing them through the `__init__` method.

Reply:
You could set these attributes inside of `__init__` but not make them arguments. This way, it's still just one line per argument.

The disadvantage of class attributes is that if they are changed on the class itself, it affects the behavior of all instances (I admit this is an edge case). Sometimes this behavior is desired, but often it comes as a surprise. Setting class attributes is getting more common since `dataclass`es were introduced, but there, the `__init__` is created implicitly, so the described problem does not occur.

Reply:
I see yes, this is a great point ! This will make everything less error-prone
Will update accordingly